File: Services/RabbitMqService.cs
using RabbitMQ.Client;
using System.Text;
using System.Text.Json;
using PaymentService.Models;

namespace PaymentService.Services;

public interface IRabbitMqService
{
    Task ConnectAsync();
    Task PublishBudgetCreatedAsync(Budget budget);
    Task PublishPaymentCompletedAsync(Payment payment, Budget budget);
    Task PublishPaymentFailedAsync(Payment payment, string reason);
    Task CloseAsync();
}

public class RabbitMqService : IRabbitMqService
{
    private IConnection? _connection;
    private IModel? _channel;
    private readonly IConfiguration _configuration;
    private readonly ILogger<RabbitMqService> _logger;

    public RabbitMqService(IConfiguration configuration, ILogger<RabbitMqService> logger)
    {
        _configuration = configuration;
        _logger = logger;
    }

    public async Task ConnectAsync()
    {
        try
        {
            await Task.Run(() =>
            {
                var factory = new ConnectionFactory
                {
                    HostName = _configuration["RabbitMq:HostName"] ?? "localhost",
                    UserName = _configuration["RabbitMq:UserName"] ?? "guest",
                    Password = _configuration["RabbitMq:Password"] ?? "guest"
                };

                _connection = factory.CreateConnection();
                _channel = _connection.CreateModel();

                // Declarar exchange
                _channel.ExchangeDeclare("payment-events", ExchangeType.Topic, durable: true, autoDelete: false);

                // Declarar filas
                _channel.QueueDeclare("budget-generated", durable: true, exclusive: false, autoDelete: false);
                _channel.QueueDeclare("payment-completed", durable: true, exclusive: false, autoDelete: false);
                _channel.QueueDeclare("payment-failed", durable: true, exclusive: false, autoDelete: false);

                // Bindings
                _channel.QueueBind("budget-generated", "payment-events", "budget.created");
                _channel.QueueBind("payment-completed", "payment-events", "payment.completed");
                _channel.QueueBind("payment-failed", "payment-events", "payment.failed");

                _logger.LogInformation("RabbitMQ conectado com sucesso");
            });
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Erro ao conectar ao RabbitMQ");
            throw;
        }
    }

    public async Task PublishBudgetCreatedAsync(Budget budget)
    {
        if (_channel == null) return;

        try
        {
            await Task.Run(() =>
            {
                var message = new
                {
                    budgetId = budget.BudgetId,
                    customerId = budget.CustomerId,
                    totalAmount = budget.TotalAmount,
                    timestamp = DateTime.UtcNow
                };

                var json = JsonSerializer.Serialize(message);
                var body = Encoding.UTF8.GetBytes(json);

                _channel.BasicPublish("payment-events", "budget.created", null, body);

                _logger.LogInformation($"Evento budget.created publicado: {budget.BudgetId}");
            });
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Erro ao publicar budget.created");
        }
    }

    public async Task PublishPaymentCompletedAsync(Payment payment, Budget budget)
    {
        if (_channel == null) return;

        try
        {
            await Task.Run(() =>
            {
                var message = new
                {
                    paymentId = payment.PaymentId,
                    budgetId = payment.BudgetId,
                    customerId = payment.CustomerId,
                    amount = payment.Amount,
                    orderId = payment.OrderId,
                    timestamp = DateTime.UtcNow
                };

                var json = JsonSerializer.Serialize(message);
                var body = Encoding.UTF8.GetBytes(json);

                _channel.BasicPublish("payment-events", "payment.completed", null, body);

                _logger.LogInformation($"Evento payment.completed publicado: {payment.PaymentId}");
            });
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Erro ao publicar payment.completed");
        }
    }

    public async Task PublishPaymentFailedAsync(Payment payment, string reason)
    {
        if (_channel == null) return;

        try
        {
            await Task.Run(() =>
            {
                var message = new
                {
                    paymentId = payment.PaymentId,
                    budgetId = payment.BudgetId,
                    customerId = payment.CustomerId,
                    amount = payment.Amount,
                    reason = reason,
                    timestamp = DateTime.UtcNow
                };

                var json = JsonSerializer.Serialize(message);
                var body = Encoding.UTF8.GetBytes(json);

                _channel.BasicPublish("payment-events", "payment.failed", null, body);

                _logger.LogInformation($"Evento payment.failed publicado: {payment.PaymentId}");
            });
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Erro ao publicar payment.failed");
        }
    }

    public async Task CloseAsync()
    {
        try
        {
            await Task.Run(() =>
            {
                _channel?.Close();
                _channel?.Dispose();
                _connection?.Close();
                _connection?.Dispose();
            });

            _logger.LogInformation("RabbitMQ desconectado com sucesso");
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Erro ao desconectar RabbitMQ");
        }
    }
}

File: src/services/RabbitMQService.js
import amqp from 'amqplib';
import config from '../config.js';

class RabbitMQService {
  constructor() {
    this.connection = null;
    this.channel = null;
  }

  async connect() {
    try {
      this.connection = await amqp.connect(config.rabbitmq.url);
      this.channel = await this.connection.createChannel();
      
      // Declarar exchanges
      await this.channel.assertExchange('payment-events', 'topic', { durable: true });
      
      // Declarar filas
      await this.channel.assertQueue('budget-generated', { durable: true });
      await this.channel.assertQueue('payment-completed', { durable: true });
      await this.channel.assertQueue('payment-failed', { durable: true });
      
      // Bindings
      await this.channel.bindQueue('budget-generated', 'payment-events', 'budget.created');
      await this.channel.bindQueue('payment-completed', 'payment-events', 'payment.completed');
      await this.channel.bindQueue('payment-failed', 'payment-events', 'payment.failed');
      
      console.log('[RabbitMQ] Conectado com sucesso');
      
      // Reconectar automaticamente
      this.connection.on('error', (err) => {
        console.error('[RabbitMQ] Erro de conexão:', err);
        setTimeout(() => this.connect(), 5000);
      });
    } catch (error) {
      console.error('[RabbitMQ] Falha ao conectar:', error);
      setTimeout(() => this.connect(), 5000);
    }
  }

  async publishBudgetCreated(budget) {
    if (!this.channel) return;
    
    try {
      this.channel.publish(
        'payment-events',
        'budget.created',
        Buffer.from(JSON.stringify({
          budgetId: budget.budgetId,
          customerId: budget.customerId,
          totalAmount: budget.totalAmount,
          timestamp: new Date().toISOString()
        }))
      );
    } catch (error) {
      console.error('[RabbitMQ] Erro ao publicar budget.created:', error);
    }
  }

  async publishPaymentCompleted(payment, budget) {
    if (!this.channel) return;
    
    try {
      this.channel.publish(
        'payment-events',
        'payment.completed',
        Buffer.from(JSON.stringify({
          paymentId: payment.paymentId,
          budgetId: payment.budgetId,
          customerId: payment.customerId,
          amount: payment.amount,
          orderId: payment.orderId,
          timestamp: new Date().toISOString()
        }))
      );
    } catch (error) {
      console.error('[RabbitMQ] Erro ao publicar payment.completed:', error);
    }
  }

  async publishPaymentFailed(payment, reason) {
    if (!this.channel) return;
    
    try {
      this.channel.publish(
        'payment-events',
        'payment.failed',
        Buffer.from(JSON.stringify({
          paymentId: payment.paymentId,
          budgetId: payment.budgetId,
          customerId: payment.customerId,
          amount: payment.amount,
          reason: reason,
          timestamp: new Date().toISOString()
        }))
      );
    } catch (error) {
      console.error('[RabbitMQ] Erro ao publicar payment.failed:', error);
    }
  }

  async consume(queueName, callback) {
    if (!this.channel) return;
    
    try {
      await this.channel.consume(queueName, async (msg) => {
        if (msg) {
          try {
            const content = JSON.parse(msg.content.toString());
            await callback(content);
            this.channel.ack(msg);
          } catch (error) {
            console.error(`[RabbitMQ] Erro ao processar mensagem da fila ${queueName}:`, error);
            this.channel.nack(msg, false, true);
          }
        }
      });
    } catch (error) {
      console.error(`[RabbitMQ] Erro ao consumir da fila ${queueName}:`, error);
    }
  }

  async close() {
    try {
      if (this.channel) await this.channel.close();
      if (this.connection) await this.connection.close();
      console.log('[RabbitMQ] Conexão fechada');
    } catch (error) {
      console.error('[RabbitMQ] Erro ao fechar conexão:', error);
    }
  }
}

export default new RabbitMQService();
